File: 2024/March/3-22 234 Palindrome Linked List/Palindrome.java

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    ListNode next;
    
    public boolean isPalindrome(ListNode head) {
        next = head;
        return recursiveResult(head);
    }

    private boolean recursiveResult(ListNode node) {
        if(node == null) {
            return true;
        }

        if(recursiveResult(node.next)) {
            if(node.val == next.val) {
                next = next.next;
                return true;
            }
            else {
                return false;
            }
        }
        else {
            return false;
        }
    }
}

File: 2024/March/3-22 234 Palindrome Linked List/Palindrome.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        def recursiveResult(node: ListNode) -> bool:
            if not node:
                return True
            
            nonlocal next
            
            if recursiveResult(node.next):
                if node.val == next.val:
                    next = next.next
                    return True
                else:
                    return False
            else:
                return False  
        
        next = head          
        return recursiveResult(head)